### ISSUE TITLE
Global packages no longer work after node.js update with homebrew

### ISSUE THREAD

**Issue by publicarray** [yarnpkg/yarn#2064]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**

When node and yarn are installed with homebrew `yarn global bin` returns a path that is tied to the a specific version of node. If node is updated with brew the symlinks are no longer valid, i.e global packages with binaries need to be reinstalled.

```bash
~ ❯ yarn global bin
/usr/local/Cellar/node/7.2.0/bin
```

Related issue: https://github.com/Homebrew/homebrew-core/issues/7283

**If the current behavior is a bug, please provide the steps to reproduce.**

I've used brew reinstall to emulate an update, I'm not sure how different it is to `brew upgrade node`.

```bash
~ ❯ brew install node
~ ❯ node --version
v7.2.0
~ ❯ brew install yarn
~ ❯ yarn global bin
/usr/local/Cellar/node/7.2.0/bin
~ ❯ yarn global add bower
~ ❯ bower --version
1.8.0
~ ❯ brew reinstall node --HEAD
~ ❯ node --version
v8.0.0-head
~ ❯ yarn global bin
/usr/local/Cellar/node/HEAD-11b0872/bin
~ ❯ bower --version
command not found: bower
```

**What is the expected behavior?**

```bash
~ ❯ bower --version
1.8.0
```

**Please mention your node.js, yarn and operating system version.**

```bash
~ ❯ yarn --version
0.17.9
~ ❯ node --version
v7.2.0
~ ❯ sw_vers
ProductName: Mac OS X
ProductVersion: 10.12.1
BuildVersion: 16B2555
```

**Comment by nicgordon** [comment]
The same issue exists for node installed via nvm:
```
~ $ which node
/Users/nicgordon/.nvm/versions/node/v7.2.0/bin/node
~ $ yarn global bin
/Users/nicgordon/.nvm/versions/node/v7.2.0/bin
~ $ yarn global add gatsby
~ $ which gatsby
gatsby not found
```

However adding this prefix flag symlinks the installed module into `/usr/local/bin` and fixes the issue for me
```
~ $ yarn global add gatsby --prefix /usr/local
~ $ which gatsby
/usr/local/bin/gatsby
```

Checking the symlink shows me that the module was actually installed here:
```
/Users/nicgordon/.config/yarn/global/node_modules/.bin/gatsby
```

Found this solution here https://github.com/yarnpkg/yarn/issues/1877

**Comment by publicarray** [comment]
Thank you, however I personally think that your fix is more of a work around and not a permanent solution. I was looking to start a discussion of how this could be handled within yarn or brew so that yarn global packages are not effected on a node.js version change. On the other hand npm global packages function fine after a `brew upgrade node`.

**Comment by kamui545** [comment]
I had the same issue several times too.
It would be nice to have this working without the prefix settings

**Comment by gutenye** [comment]
same issue here, is there a permanent solution?

**Comment by OmgImAlexis** [comment]
Got the same issue here.

**Comment by aniddan** [comment]
same issue

**Comment by bradley-curran** [comment]
Yarn adds a symlink of the executables to node's `bin` directory.
In my instance it was `/usr/local/Cellar/node/7.2.1`. 
The actual place where the global packages are installed is `~/.config/yarn/global/node_modules/`. 

When I updated to node 7.4.0 from 7.2.1 the symlinks weren't added to the new version. I'm guessing you're seeing the same thing. 

The problem is, there's no way of using yarn right now to fix this. 

You could write a script that does something along the lines of: 

- Get the global packages directory for yarn
- Find all the files that are executable
- Symlink all the files into your current node directory

I'll write an npm package later that does all this.

**Comment by bradley-curran** [comment]
Hey guys, 

I've written an npm package that should fix your yarn-installed global packages. 

```bash
$ npm install -g fix-yarn-global-packages
$ fix-yarn-global-packages
```

That should fix all your global packages. Let me know if it doesn't work for you. 

https://www.npmjs.com/package/fix-yarn-global-packages

**Comment by publicarray** [comment]
Thanks @bradley-curran but I personally prefer to just added `alias yarn-global'=yarn global --prefix /usr/local'` to my dotfiles until this is fixed.

**Comment by bradley-curran** [comment]
I've been doing some more digging and it seems the issue is that yarn doesn't read all npm config files that npm itself uses. 

Yarn should read all files defined in [npm's npmrc documentation](https://docs.npmjs.com/files/npmrc)

[Currently yarn only reads config files](https://github.com/yarnpkg/yarn/blob/67cc334e5303c314f7a1949a7ddd2d77f66eb4f3/src/registries/npm-registry.js#L98) from: 

- `./.npmrc`
- `~/.npmrc`
- `<node directory>/.npmrc` (\<node directory\> = `/usr/local/Cellar/node/7.6.0/` for me)

The config files that also need to be read include: 

- `$PREFIX/etc/npmrc` ($PREFIX is usually `/usr/local`)
- `path/to/npm/itself/npmrc`

The problem is that these files can't be found easily using the existing architecture afaict. 
Ideally the existing config loading would be replaced with delegating to `npm config ls -l` and parsing the result rather than trying to read all the files. 

### Quick Fix

- Open `~/.npmrc` in your favourite text editor. 
- Add the line `prefix = /usr/local`

### The Real Fix
Read all configuration files, ideally by delegating to `npm config`, rather than reading existing INI files. I'd be happy to contribute, though I'm not really sure who I can discuss implementation details with.

**Comment by publicarray** [comment]
@bradley-curran thanks so much for looking into this, from now on I'll use the npmrc file. (didn't know that the npmrc existed until now). Just a small word of warning that npm will also install global packages to prefix set in in the npmrc and no longer in `/usr/local/Cellar/node/7.6.0/` as previously.

**Comment by mkutny** [comment]
Agree it should not install into /usr/local/Cellar/node/version .

As a temporary fix, it's now possible to `yarn config set prefix /usr/local -g`

Although this prefix should be set by default on yarn installation.

**Comment by publicarray** [comment]
Similar issue: https://github.com/yarnpkg/yarn/issues/2639

**Comment by publicarray** [comment]
@mkutny is the `-g` flag supposed to do something? for me the config file that is modified is always `~/.yarnrc`

**Comment by mkutny** [comment]
@publicarray , I just did it according to the doc which says `yarn config set <key> <value> [-g|--global] `

It seems that in this particular case it doesn't matter but for options which could be set for either local or global scope, it does.

**Comment by publicarray** [comment]
@mkutny thanks, I should have read the docs 📖

**Comment by HenrikWL** [comment]
Can't this be fixed by having the homebrew formula do

`yarn config set prefix /usr/local/`

as a post-install?

**Comment by bradley-curran** [comment]
You can, if you're willing to assume that the prefix is /usr/local on all machines, which I'm not sure can be assumed.

**Comment by HenrikWL** [comment]
Good point. Within the context of Homebrew formulae, I do believe the global Homebrew-prefix is available in a variable. It would be better to use that.

**Comment by vdh** [comment]
Please stop advocating the `/usr/local` prefix workaround. Some of us prefer to stick with the versioned bin directory (e.g. `/usr/local/Cellar/node/7.9.0/bin`) and keep `/usr/local/bin` clean. Editing `$PATH` is a whole lot easier than cleaning up dead symlinks from an overloaded `/usr/local/bin` directory.

**Comment by mkutny** [comment]
@vdh , it's not about which solution "is a whole lot easier". It's all about what is the right thing to do.

If you read the thread carefully there are currently several issues with global packages:
- they are not immediately available after `yarn global add` (they need to be manually symlinked)
- global packages won't work after node update (`$PATH` needs to be updated)
- global packages will be removed if obsolete node version is purged (editing `$PATH` won't help at all)

Binaries in `/usr/local/bin` not only solve all these issues but are also consistent with Unix way of doing things in general and with Homebrew in particular.

**Comment by HenrikWL** [comment]
@vdh the solution to dead symlinks in /usr/local/bin is to fix the problem that causes dead symlinks in there, not to preserve broken behaviour. Like @mkutny says, there are several issues with `yarn global` the way it works now. I expect it to work like `npm -g` but it doesn't. If that's by design then that's ok, I'll just stick with `npm -g` for global packages and keep using yarn for project packages (for which it works great).

**Comment by partounian** [comment]
I'm not sure if this is related or not, but with zsh, I have an export on startup with the path from `yarn global bin`, but on startup I always get `command not found: yarn` how it calling yarn does work. On top of this with global packages that have their own command, I get command not found when trying that call them, but it all works fine when I reload zsh in the same terminal window.

**Comment by publicarray** [comment]
@partounian I recommend you open a new issue because your problem is unrelated. Is it possible that you are calling yarn before you initialised the $PATH? e.g.

~/.zshrc:
```sh
yarn global bin
export PATH=$(brew --prefix)/bin:$PATH # run this first
```

**Comment by partounian** [comment]
Thanks, I found that my dotfiles includes did not take into account dependencies.

**Comment by vdh** [comment]
@mkutny 

>  it's not about which solution "is a whole lot easier". It's all about what is the right thing to do.

Yes, and the right thing to do is to use `$PATH`. It's a fundamental underpinning of Unix. Version management tools like NVM and RVM are designed around automating changes to `$PATH`. On macOS, `/usr/local` is much more sane when managed by Homebrew. I'm sure Linux has similar conventions with their package mangers as well.

`/usr/local/bin` should be reserved for real binaries and symlinks (ideally, managed by your package manager of choice), not Node scripts. Anything installed by Yarn/NPM are (usually) not real binaries, they're scripts that rely on the `node` binary to function. It makes logical sense to keep them segregated and located with their dependency, in the same way I don't throw all my global binaries/scripts from other languages like Ruby/Python/etc into `/usr/local/bin` either.

> they are not immediately available after yarn global add (they need to be manually symlinked)

You only have that issue because you don't add `/usr/local/Cellar/node/7.9.0/bin` (or equivalent Node version) to your `$PATH`. You're creating your own extra work if you decide you want to manage via symlinks instead of managing via the `$PATH`.

> global packages won't work after node update ($PATH needs to be updated)

You're not updating your global packages, you're just passing them through by the coincidence of their location staying the same, and hoping nothing breaks. `node-sass` in particular usually needs a rebuild for its C binaries after most Node updates. It's less of a headache to rebuild all global packages after a Node update, instead of just assuming all your Node scripts will be compatible with a newer binary without making any checks.

> global packages will be removed if obsolete node version is purged (editing $PATH won't help at all)

That's a **feature** for me, I don't want outdated scripts hanging around. Each new version of the Node binary should be a fresh slate to reinstall your global packages against. As much as authors will try to make their scripts cross-compatible, *everything* eventually has a Node version dependency.

 @HenrikWL 
Yarn should behave better than NPM. Installing into `/usr/local` should be a choice made by the user who wants to that kind of setup, but it shouldn't be enforced or setup that way by default. 

IMO, installing into the Node `bin` is better behaviour than NPM's `/usr/local` default. Live by the Node binary, die by the Node binary.

**Comment by zenopopovici** [comment]
@vdh Sorry, but it's ridiculous to suggest a re-install all global packages with every new node version. Some of us have 20+ packages that take bandwidth and time to reinstall, and all of them worked fine for all major versions of node updates. It's not a _feature_ it's a _bug_.

Just manage for now your global packages with `npm` until they figure this out. More over Yarn doesn't have an `outdated` for global packages yet.

**Comment by publicarray** [comment]
@vdh I appreciate that yarn works for you just the way it is. For me I don't particularly care much about the implementation details but whatever the behavior it should be documented (to my knowledge the current behavior is not documented). So at the moment with the default behaviour global packages break in a bad way when I do a routine `brew upgrade`.

Most of the packages that I use are mostly written in JavaScript and will function fine in the future e.g. eslint, prettier, babel, fast, lighthouse etc... and like @zenopopovici said I don't want to reinstall them every time a node is updated. This issue highlights the need to allow the user to have global packages that function across node version or at the very least it should be made absolutely clear that yarn has a different behavior than npm. Unfortunately yarn is more or less marketed as a npm replacement.

Setting the prefix to `/usr/local` or using npm for globals is only a temporary workaround until a solution is found.

**Comment by mkutny** [comment]
@vdh ,
I see you have some difficulties with installing executables in `/usr/local/bin` but given your explanation, it's hard for me to realize what are they exactly.

Putting executables in `/usr/local/bin` is the "right way to do" not only because it's a de-facto standard for *nix-like systems but also it's a real standard (it's a part of POSIX).

When you do it in a non-standard way you break not only people expectations but also POSIX-compliant (read almost all) utilities/packages/software that depend on `/usr/local/bin`.

As you may see from this thread (and a whole bunch of similar/duplicated `yarn` issues) a lot of people have a lot of pain because of `yarn`'s non-standard approach to handling binaries.

BTW, *nix systems don't distinguish between "binaries" and "scripts". What they care if a file is executable or not. And if a "script" depends on some interpreter (ruby/python/whatever) it's a task of a package manager to handle that dependency by installing the interpreter. It's all could be done by obeying the standard, not going against it.

**Comment by HenrikWL** [comment]
And furthermore, if the interpreter and the stuff running on it is so fragile that it risks breaking at any given update of the interpreter then that's a sign of an unhealthy ecosystem.

That being said, I've never had anything break due to node upgrades, except for cases where our project had fell into disrepair for years. Is this actually such an issue that we need to invent these convoluted workarounds?

Most of the stuff I install as global node modules are things like firebase-tools, bower, grunt-cli and such - these are general purpose utilities, and for them to be mapped 1-1 to individual node versions is completely unfathomable to me.

**Comment by vdh** [comment]
@zenopopovici 

Yarn caches downloads, no bandwidth will be wasted. If Yarn decides you need a new version it will fetch it, otherwise it'll reuse the existing cache. Also, nobody is going to stop you from setting a prefix of `/usr/local` if you want your own setup to reuse existing builds the way NPM does.

@mkutny

> I see you have some difficulties with installing executables in `/usr/local/bin`

"Difficulties"? Absolutely not. It's a conscious choice to not throw everything into one directory. It's a very common pattern in the *nix world to add additional `bin` directories to the `$PATH` variable. X11/XQuartz does it. Multiple library frameworks do it (Java, Mono, etc). Take a look at `/etc/paths` and (if it exists in your OS) `/etc/paths.d`. In your own words, "it's a part of POSIX".

It's standard, please don't make wild accusations about "non-standard" and breaking things. Throwing everything into `/usr/local` is a **choice**, not a "standard". NPM behaves that way to be easy on newcomers who are inexperienced at configuring their environment settings. But "zero config" design choices does not automatically make it the "right" thing to do.

Heck, I don't even have that many global scripts to begin with, because [Direnv](https://direnv.net/) is a great little tool to automate `$PATH` changes per-directory, dynamically referencing project-specific `node_modules/.bin`

@HenrikWL 

> And furthermore, if the interpreter and the stuff running on it is so fragile that it risks breaking at any given update of the interpreter then that's a sign of an unhealthy ecosystem.

Is there a different Javascript ecosystem out there somewhere that I'm not aware of…? One that *doesn't* contain massive amounts of churn…? Look, I *like* Node, but JS is quite frankly **insane** compared to any other scripting language's ecosystem.

> That being said, I've never had anything break due to node upgrades, except for cases where our project had fell into disrepair for years. Is this actually such an issue that we need to invent these convoluted workarounds?

"Works on my machine" is an unhealthy mindset to have, it's disingenuous to dismiss problems just because you haven't experienced them yourself.

---

**TLDR:** To each their own if you want to throw everything into `/usr/local`, but don't force NPM's design decisions onto Yarn. Yarn isn't "broken", please don't malign it for using a mechanism you find unfamiliar. You're one prefix setting away from having NPM's behaviour if you want it that way, or you can just use NPM.

**Comment by publicarray** [comment]
@vdh We are not forcing anything we just want better behavior with brew/nvm. You have a certain way you work with a tool and that's fine. I have different needs than you do and since yarn is marketed as a faster replacement than npm it should be no surprise that this issue exists. if the tool does not support global packages like I expect that is fine. Just make it clear before I install it, update my brew packages and break all of my global yarn packages. (The first time that happened it took me hours to reasonably understand what was going on and a few more hours to fix it). The [docs literally](https://yarnpkg.com/en/docs/migrating-from-npm) say that `npm install --global [package]` = `yarn global add [package]` 

Also not everyone shares your workflow please respect that.

**Comment by vdh** [comment]
@publicarray 
Yarn is a drop-in replacement that makes different design choices than NPM. It shouldn't be expected to replicate exact behaviour. The literal title is "CLI commands comparison", and some are *(N/A)* because they differ in design. How is asking the default prefix to change to be more like NPM *not* forcing things?

**Comment by publicarray** [comment]
> Yarn is a drop-in replacement that makes different design choices than NPM. It shouldn't be expected to replicate exact behaviour. The literal title is "CLI commands comparison", and some are (N/A) because they differ in design.

By that logic `npm install --global [package]` should = N/A. but it doesn't.

> How is asking the default prefix to change to be more like NPM not forcing things?

That was never the intention. Please re-read the entire issue.

From my comment above: 
> Setting the prefix to /usr/local or using npm for globals is only a temporary workaround until a solution is found.

**Comment by publicarray** [comment]
The ideal solution is to make yarn read a global config file that brew can set. So that the brew package manager can set the prefix it defines e.g. /usr/local/yarn

**Comment by vdh** [comment]
@publicarray Yarn *already* installs into the same place as Homebrew. And Yarn can already be configured to install to a custom location. These arguments aren't making sense.

**Comment by luftywiranda13** [comment]
yes i'm exactly the same problem.
Globally installed packages are working just fine

But yarn global ls shows nothing instead of just done

**Comment by mkutny** [comment]
> >I see you have some difficulties with installing executables in `/usr/local/bin`
> 
> "Difficulties"? Absolutely not. It's a conscious choice to not throw everything into one directory. It's a very common pattern in the *nix world to add additional bin directories to the $PATH variable.

The `PATH` variable is not a replacement for the organized filesystem. The whole purpose of `PATH` is to facilitate less typing so instead of `/usr/local/bower` you go just `bower`.

Executables SHOULD go to `/bin`, `/usr/bin`, `/usr/local/bin`, `/usr/X11R6/bin` etc and you MAY point `PATH` to them.

`brew` (Mac), `dpkg` (Debian), `rpm` (RedHat), `npm` - all package managers work this way.

Technically you can put an executable into `/whatever/directory/i/want` directory but (whether you point `PATH` there or not) it is non-standard and goes against people's expectations.

It's absolutely legit to keep several versions of `node` and switch between them. Package managers achieve this by maintaining a symbolic link in `/usr/local/bin` determining default version. If you want to achieve a similar effect by manipulating the `PATH` variable you are absolutely free to do it. Just mind that placing symlinks in `*/bin` does not stand in your way.

**Comment by vdh** [comment]
@mkutny What is all this "non-standard" and "goes against people's expectations" you keep throwing around?

Everything inside `/usr/local` is dictated by *you*, the user. There are no other people who's expectations matter to that directory except your own, unless you're running a multi-user machine (e.g. a server). Homebrew has built up its own *conventions* of how they want to structure their directories and manage symlinks, but that is all just their convention. If you want to throw around the word "standard", please quote an official (or semi-official) source.

Here, let's quote [Homebrew's FAQ page](http://docs.brew.sh/FAQ.html), since "standard Homebrew" is the closest thing we're going to get to a "standard" right now:

> ## Why does Homebrew prefer I install to /usr/local? ##
> 1) It’s easier
> /usr/local/bin is already in your PATH.
> 2) It’s easier
> Tons of build scripts break if their dependencies aren’t in either /usr or /usr/local. We fix this for > Homebrew formulae (although we don’t always test for it), but you’ll find that many RubyGems > and Python setup scripts break which is something outside our control.
> 3) It’s safe
> Apple has left this directory for us. Which means there is no /usr/local directory by default, so > there is no need to worry about messing up existing tools.
> If you plan to install gems that depend on brews then save yourself a bunch of hassle and install to /usr/local!
>
> It is not always straightforward to tell gem to look in non-standard directories for headers and libraries. If you choose /usr/local, many things will “just work”.

and further down:

> ## Can I install my own stuff to /usr/local? ##
> Yes, brew is designed to not get in your way so you can use it how you like.
>
> Install your own stuff, but be aware that if you install common libraries like libexpat yourself, it may cause trouble when trying to build certain Homebrew formula. As a result brew doctor will warn you about this.
>
> Thus it’s probably better to install your own stuff to the Cellar and then brew link it. Like so: [...]

So yes, it certainly is *easier* to put your binaries into `/usr/local/bin`, *if* you don't know how to (or care to) add new paths to `PATH`. But it's all pure convention, everything in there (at least on macOS) is Homebrew's own structure. Pointing NPM/Yarn's symlinks into `/usr/local/bin` is purely a *choice* by someone who does not want to modify their `PATH`.

You can run `yarn config set prefix /usr/local -g` and set your Yarn prefix there as well, for the method *you* find easier. Myself, I find it much *much* easier to set a config in `/etc/paths.d` (and as the FAQ points out, this wont generate warnings from `brew doctor`).

Let's all stop pretending our own *personal* preferences makes someone else's preferences "non-standard".

**Comment by mkutny** [comment]
> What is all this "non-standard" and "goes against people's expectations" you keep throwing around?
> If you want to throw around the word "standard", please quote an official (or semi-official) source.

I quoted it earlier but you apparently missed it. It is called Filesystem Hierarchy Standard and it's a part of POSIX. In particular, following [this link to spec](http://refspecs.linuxfoundation.org/FHS_2.3/fhs-2.3.html#USRLOCALLOCALHIERARCHY) you'll find what it says about `/usr/local`.


> Homebrew has built up its own _conventions_ of how they want to structure their directories and manage symlinks, but that is all just their convention.

Note, that their "own" conventions just follow FHS (surprise). Also note, that standards **are** conventions. Look, people just say: "There are a lot of different ways of doing things. Let's pick and agree on **one** of such ways so it will be easier for us to cope with each other". One of such conventions is FHS and Homebrew just follows it.


> Myself, I find it much _much_ easier to set a config in `/etc/paths.d`.

That's absolutely fine! What we propose is to put symlinks into `/usr/local/bin` which not only follow conventions but also does not break your own way of doing things. Unfortunately what you propose just breaks ours.

**Comment by vdh** [comment]
@mkutny 
Following FHS doesn't say anywhere there to throw every single one of your symlinks into `/usr/local/bin`. It doesn't even say whether to put symlinks in there vs the actual binaries. FHS just dictates common directory structures.

>  but also does not break your own way of doing things

Yes it *does*, Homebrew manages symlinks inside `/usr/local/bin` just fine on its own. I don't need tons of non-Homebrew symlinks in there for a management headache and for `brew doctor` to warn about.

*I'm* not the one proposing a break in behaviour. Yarn is working fine the way it was designed. A developer implemented this current behaviour for a reason. You're the ones proposing a breaking change just to make it more like NPM's default behaviour. Once you start forcing those symlinks to mix into `/usr/local/bin` it will be a huge pain trying to un-mix them after the fact.

What is it about setting `yarn config set prefix /usr/local -g` once that offends you so much that you want to force your breaking change upon Yarn?

**Comment by mkutny** [comment]
> I don't need tons of non-Homebrew symlinks in there for a management headache and for `brew doctor` to warn about.

My `brew doctor` does not warn me about non-Homebrew symlinks.
But even if it did... here's what it would print out:

```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!
```

Homebrew maintainers suggest you to not worry. You might not like non-Homebrew symlinks in `/usr/local/bin` but they are _not breaking_.
Homebrew is designed to not get in your way to work with `/usr/local`. It can manage symlinks either on its own or side by side with another package manager.


> Yarn is working fine the way it was designed.

A lot of people here sure it's not. Just an excerpt from a complaint above the thread:
```
~ $ yarn global add gatsby
~ $ which gatsby
gatsby not found
```

I can hardly agree it looks like "working fine". For me **gatsby not found** means "not working at all and needs to be fixed".

> What is it about setting `yarn config set prefix /usr/local -g`?

That's exactly what we propose to do here!

**Comment by HenrikWL** [comment]
I'd be happy if the prefix option worked. However, it only partially does:

![skjermbilde 2017-04-26 kl 10 20 47](https://cloud.githubusercontent.com/assets/116110/25424847/2a0a89a6-2a6a-11e7-9e12-99416a71f048.png)

**Comment by mkutny** [comment]
@HenrikWL , it worked in v0.21 and now it's broken (but I think it's a separate issue).

**Comment by HenrikWL** [comment]
@mkutny yeah, this thread has kinda bloated a bit, I think there's at least 3 different issues being discussed here. If there's a dedicated one for `yarn global ls` not working when the prefix option is set, I'll hit that one up instead.

**Comment by mkutny** [comment]
@HenrikWL, issues with `yarn global ls` have been reported a while ago, they been being fixed in PR-3238.

**Comment by Daniel15** [comment]
Adding `~/.config/yarn/global/node_modules/.bin` to your `$PATH` is probably a reasonable workaround for now. 

The issue is that Yarn symlinks the executables into the wrong directory (the Node.js directory rather than somewhere reasonable). It doesn't really make sense to use the Node.js directory for these scripts, and on many systems that directory is not even writable by regular users. I fixed this for Windows in https://github.com/yarnpkg/yarn/pull/3233 but we still need fixes for other operating systems. https://github.com/yarnpkg/yarn/pull/3458 seems reasonable as long as the user has write access to `/usr/local/bin`, although maybe something like `~/.yarn/bin` would be better.

**Comment by KishanBagaria** [comment]
**Windows and macOS**
`/usr/local/bin` is writable in a default macOS installation and that's where Homebrew writes stuff. When node/npm is installed from Homebrew, [Homebrew edits the prefix config in `.npmrc` so that npm links the global binaries in `/usr/local/bin` too](https://github.com/Homebrew/homebrew-core/blob/master/Formula/node.rb#L115).

For Windows and macOS, we can use `%LocalAppData%\Yarn\bin` and `/usr/local/bin` as default, because that's guaranteed to work out-of-the-box (unless the user is on a broken/unusual installation).

On macOS, Yarn global binaries will work without extending the `PATH` since `/usr/local/bin` is present in default `PATH`. Although, the directories in default Windows `PATH` all need admin privileges, so Chocolatey or the user will have to extend the `PATH` with `%LocalAppData%\Yarn\bin`.

**Linux et al.**
For Linux and other unrecognized platforms, we shouldn't assume anything (any of the parent directories of `process.execPath` isn't reasonable or guaranteed to be writable as @Daniel15 [suggested](https://github.com/yarnpkg/yarn/issues/2064#issuecomment-303592899)). We can test if `/usr/local/bin` is writable before using it. But there's thousands of configurations of Linux distros with different versions and different defaults. I dunno if that directory will work on all of them.
Instead, we can ask the user to set the prefix explicitly. Running any `yarn global` command on these platforms should throw an error and say something like
> *error* you need to first run `yarn config set prefix <global_prefix_dir>`
> *info* see https://yarnpkg.com/en/docs/cli/global

Or we can show an interactive `inquirer.js` prompt where the user can select one among many commonly-used prefixes:
```
* /usr/local/bin
* /usr/bin
* /bin
* ~/n/bin
* ~/bin
* <enter custom prefix>: ______
```

The current behaviour has been copied from npm which was written about 5 years ago: https://github.com/npm/npm/blob/771c66884a0a1851cffb3bf132828f6ccef1c09c/lib/config/defaults.js#L92-L105

**Comment by BYK** [comment]
>When node/npm is installed from Homebrew, Homebrew edits the prefix config in .npmrc so that npm links the global binaries in /usr/local/bin too.

Shall we honor this setting too?

>Although, the directories in default Windows PATH all need admin privileges, so Chocolatey or the user will have to extend the PATH with %LocalAppData%\Yarn\bin.

I think this is a really subpar experience for Windows users. Node puts itself into the path along with npm right, why can't we use those directories directly?

**Comment by Daniel15** [comment]
> I think this is a really subpar experience for Windows users

The Windows installer (and Chocolatey package) automatically update the path, so this is mostly a non-issue. 

> Node puts itself into the path along with npm right, why can't we use those directories directly?

On most systems, it would require you to run Yarn as administrator, as Node is normally installed at C:\Program Files\nodejs and Node.js doesn't expose the Win32 API for requesting elevation (getting admin rights after the app has started, like "sudo")

**Comment by BYK** [comment]
@Daniel15 oh, okay so this clears my Windows-side questions :)

**Comment by BYK** [comment]
Well, with the exception of, when/if we install `yarn` via NPM or as a direct script. I'm not sure if we need elevated permissions when running `npm install -g` do we?

**Comment by Daniel15** [comment]
> I'm not sure if we need elevated permissions when running `npm install -g` do we?

I think it'd depend on where Node.js is installed and what the permissions are on that directory.

**Comment by Daniel15** [comment]
> Well, with the exception of, when/if we install `yarn` via NPM or as a direct script.

We could probably update Yarn so that it edits the path on first run. However I don't know if Node.js exposes an API for that. It's got extremely limited ffi capabilities.

**Comment by BYK** [comment]
Did some reading on https://docs.npmjs.com/files/folders#prefix-configuration and looks like the approach in the diff is correct. Just concerned about not using the `npm` folder (we use `Yarn` instead). I think this is fine until we get complaints.

**Comment by nodox** [comment]
`sudo yarn global add ...` worked but I'm not sure that should be the best solution

**Comment by diegoazh** [comment]
I had the same issue in ubuntu after upgrade node version with `nvm`, in my dev proyects I fix it removing `.lock` file en running `yarn cache clean` and reinstall its. I think that if you remove the `.lock` file inside of `~/.config/yarn/global/` (in ubuntu) and reinstall all packages fix the error with global packages; also I recomend defining install location with `yarn config set prefix ~/.yarn` like [this docs](https://yarnpkg.com/lang/en/docs/cli/global/) suggests
